src/Examples/Amortized.agda

```diff
@@ -245,238 +234,137 @@ module Queue where
-    step-ret-injective : (c₁ c₂ : ℂ) (v₁ v₂ : val A) →
-      step (F A) c₁ (ret v₁) ≡ step (F A) c₂ (ret v₂) → v₁ ≡ v₂
+  _≈'_ : (q₁ q₂ : cmp (queue D)) → Set
+  _≈'_ q₁ q₂ = (A : tp pos) (p : val (queue-program A)) → ψ p q₁ ≡ ψ p q₂
 
   {-# TERMINATING #-}
-  classic-amortization :
-    (∀ {A} e → Σ ℂ λ c → Σ (val A) λ v → e ≡ step (F A) c (ret v))
-    → (∀ {A B c e f} → bind {A = A} (F B) e (step (F B) c ∘ f) ≡ step (F B) c (bind (F B) e f))
-    → (q₁ q₂ : cmp (queue (F unit)))
-    → q₁ ≈ q₂ ⇔ (∀ A → (p : val (queue-program E A)) → ψ p q₁ ≡ ψ p q₂)
-  classic-amortization writer bind/step/commutative q₁ q₂ = record
-    { f = forward q₁ q₂
-    ; g = backward q₁ q₂
-    ; cong₁ = Eq.cong (forward q₁ q₂)
-    ; cong₂ = Eq.cong (backward q₁ q₂)
+  classic-amortization : {q₁ q₂ : cmp (queue D)} → q₁ ≈ q₂ ⇔ q₁ ≈' q₂
+  classic-amortization = record
+    { f = forward
+    ; g = backward
+    ; cong₁ = Eq.cong forward
+    ; cong₂ = Eq.cong backward
     }
     where
-      forward : (q₁ q₂ : cmp (queue (F unit))) →
-        q₁ ≈ q₂ → (∀ A → (p : val (queue-program E A)) → ψ p q₁ ≡ ψ p q₂)
-      forward q₁ q₂ h A (return x) =
-        Eq.cong (λ e → bind (F A) e (λ _ → ret x)) (_≈_.quit h)
-      forward q₁ q₂ h A (enqueue e p) = forward (Queue.enqueue q₁ e) (Queue.enqueue q₂ e) (_≈_.enqueue h e) A p
-      forward q₁ q₂ h A (dequeue f) with _≈_.dequeue h
-      ... | c₁ , e , q₁' , h₁ , c₂ , _ , q₂' , h₂ , refl , hq' =
-        let open ≡-Reasoning in
-        begin
-          (bind (F A) (Queue.dequeue q₁) λ (e , q₁') →
-          bind (F A) (f e) λ p' →
-          ψ p' q₁')
-        ≡⟨ Eq.cong (λ e → bind (F A) e λ (e , q₁') → bind (F A) (f e) λ p' → ψ p' q₁') h₁ ⟩
-          (bind (F A) (step (F (prod⁺ (maybe E) (U (queue (F unit))))) c₁ (ret (e , q₁'))) λ (e , q₁') →
-          bind (F A) (f e) λ p' →
-          ψ p' q₁')
-        ≡⟨⟩
-          step (F A) c₁ (
-          bind (F A) (f e) λ p' →
-          ψ p' q₁')
-        ≡˘⟨ bind/step/commutative {B = A} {c = c₁} {e = f e} ⟩
-          (bind (F A) (f e) λ p' →
-          step (F A) c₁ (ψ p' q₁'))
-        ≡⟨
-          Eq.cong (bind (F A) (f e)) (funext λ p' →
+      forward : {q₁ q₂ : cmp (queue D)} → q₁ ≈ q₂ → q₁ ≈' q₂
+      forward h A (return x) =
+        Eq.cong (λ e → bind (F A) (out e) (const (ret x))) (_≈_.quit h)
+      forward h A (enqueue e p) =
+        forward (_≈_.enqueue h e) A p
+      forward h A (dequeue f) =
+        Eq.cong₂
+          (λ e₁ e₂ → bind (F A) (f e₁) e₂)
+          (funext/Ω λ u → proj₁ (_≈_.dequeue h) u)
+          (funext (forward (proj₂ (_≈_.dequeue h)) A))
+
+      postulate
+        ret-injective : ∀ {x y} → ret x ≡ ret y → x ≡ y
+
+      backward : {q₁ q₂ : cmp (queue D)} → q₁ ≈' q₂ → q₁ ≈ q₂
+      _≈_.quit (backward typical) = extension-≡ (typical unit (return triv))
+      _≈_.enqueue (backward typical) e =
+        backward (λ A p → typical A (enqueue e p))
+      _≈_.dequeue (backward {q₁} {q₂} typical) =
+        (λ u → ret-injective $
```

Comment:
This is the only use of `u` from `ext/val`/`◯⁺`; the rest are constant functions. Likely means that it should be removed/this proof case should change?